### PR TITLE
Extend `qcut.tape_to_graph` to support sampling measurements

### PR DESCRIPTION
**Pull request by anthayes92**
**Context:**
Many practical applications of quantum circuits require sampling measurements. Here we begin extending the existing manual circuit cutting pipeline to support sampling measurements.

**Description of the Change:**
The existing `qcut.tape_to_graph` method has been extended to convert a sampling in the computational basis over all wires in a circuit. The single sampling measurement is converted into distinct nodes in the computational graph; one for each wire. 

**Benefits:**
First step toward supporting cutting circuit with sampling measurements.

**Possible Drawbacks:**
~~Does not support sample measurements containing observables.~~

**Related GitHub Issues:**
#2297 sampling with and without observables in a single circuit causes an error. As such, the sample nodes added to the graph contain an explicit projection measurement equivalent to sampling in the computational basis. This is required for later steps in the cutting pipeline when converting subgraphs to fragment tapes.

### CODE REVIEW

File: pennylane/transforms/qcut.py [hunk 1]
```diff
@@ -239,7 +239,10 @@ def tape_to_graph(tape: QuantumTape) -> MultiDiGraph:
                 m_ = MeasurementProcess(m.return_type, obs=o)
```
Comment (trbromley) [inline pennylane/transforms/qcut.py:239]:
```python
            if m.return_type is Sample:
                 raise ValueError("Sampling from tensor products of observables "
                                           "is not supported in circuit cutting")
            for o in obs.obs:
```
Suggestion for line 238

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -4,6 +4,13 @@
 
 <h3>New features since last release</h3>
 
+* Development of a circuit-cutting compiler extension to circuits with sampling
+  measurements has begun:
+
+  - The existing `qcut.tape_to_graph()` method has been extended to convert a
+    sample measurements over all circuit wires to distinct sample nodes for each wire
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:11]:
```suggestion
    sample measurement without an observable specified to multiple single-qubit sample
    nodes.
```

File: pennylane/transforms/qcut.py (hunk 1 above)
Comment (trbromley) [reply pennylane/transforms/qcut.py:239]: \+ will need a test

File: tests/transforms/test_qcut.py [hunk 3]
```diff
@@ -382,6 +382,37 @@ def test_multiple_observables(self):
...
+        correctly converted to a graph with a distinct node for each wire sampled
+        """
+
+        with qml.tape.QuantumTape() as tape:
+            qml.Hadamard(wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.PauliX(wires=1)
+            qml.WireCut(wires=1)
+            qml.CNOT(wires=[1, 2])
+            qml.sample(wires=[0, 1, 2])
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:397]: Can we test similarly for the circuit:
```python
        with qml.tape.QuantumTape() as tape:
            qml.Hadamard(wires=0)
            qml.CNOT(wires=[0, 1])
            qml.PauliX(wires=1)
            qml.WireCut(wires=1)
            qml.CNOT(wires=[1, 2])
            qml.sample(qml.PauliZ(0))
            qml.sample(qml.PauliZ(1))
            qml.sample(qml.PauliZ(2))
```
i.e., that the nodes are
```python
        expected_nodes = [
            qml.Hadamard(wires=[0]),
            qml.CNOT(wires=[0, 1]),
            qml.PauliX(wires=[1]),
            qml.WireCut(wires=[1]),
            qml.CNOT(wires=[1, 2]),
            qml.sample(qml.PauliZ(0)),
            qml.sample(qml.PauliZ(1)),
            qml.sample(qml.PauliZ(2)),
        ]
```
I know we are thinking to restrict to only allow users to specify `qml.sample(wires=[...])` when circuit cutting, but I think the `tape_to_graph` function can already support either choice.

**Review by trbromley — CHANGES_REQUESTED** [review]
Thanks @anthayes92! Looks good, left a few suggestions.

File: tests/transforms/test_qcut.py (hunk 3 above)
Comment (anthayes92) [reply tests/transforms/test_qcut.py:397]: Yes you're right, it is supported! Added: 5a4f9e7146c61d1248c0b8dc75310d36226d8988

File: tests/transforms/test_qcut.py [hunk 4]
```diff
@@ -382,6 +382,87 @@ def test_multiple_observables(self):
             assert node_obs.wires == exp_obs.wires
             assert node_obs.obs.name == exp_obs.obs.name
 
+    def test_split_sample_measurement(self):
+        """
+        Test that a circuit with single sample measurement over all wires is
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:387]:
```suggestion
        Test that a circuit with a single sample measurement over all wires is
```

File: tests/transforms/test_qcut.py [hunk 5]
```diff
@@ -382,6 +382,87 @@ def test_multiple_observables(self):
...
+            qml.WireCut(wires=[1]),
+            qml.CNOT(wires=[1, 2]),
+            qml.sample(qml.Projector([1], wires=[0])),
+            qml.sample(qml.Projector([1], wires=[1])),
+            qml.sample(qml.Projector([1], wires=[2])),
+        ]
+
+        for node, expected_node in zip(list(g.nodes), expected_nodes):
+            assert node.name == expected_node.name
+            assert node.wires == expected_node.wires
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:414]:
```suggestion
            assert node.wires == expected_node.wires

            if getattr(node, "obs", None) is not None:
                assert node.return_type is qml.operation.Sample
                assert node.obs.name == expected_node.obs.name
```
For measurement nodes, their name is simply `"Identity"`.

File: tests/transforms/test_qcut.py [hunk 6]
```diff
@@ -382,6 +382,87 @@ def test_multiple_observables(self):
...
+            qml.WireCut(wires=[1]),
+            qml.CNOT(wires=[1, 2]),
+            qml.sample(qml.PauliZ(wires=[0])),
+            qml.sample(qml.PauliZ(wires=[1])),
+            qml.sample(qml.PauliZ(wires=[2])),
+        ]
+
+        for node, expected_node in zip(list(g.nodes), expected_nodes):
+            assert node.name == expected_node.name
+            assert node.wires == expected_node.wires
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:464]:
```suggestion
            assert node.wires == expected_node.wires

            if getattr(node, "obs", None) is not None:
                assert node.return_type is qml.operation.Sample
                assert node.obs.name == expected_node.obs.name
```

File: tests/transforms/test_qcut.py [hunk 7]
```diff
@@ -382,6 +382,87 @@ def test_multiple_observables(self):
...
+            qml.CNOT(wires=[0, 1]),
+            qml.PauliX(wires=[1]),
+            qml.WireCut(wires=[1]),
+            qml.CNOT(wires=[1, 2]),
+            qml.sample(qml.Projector([1], wires=[0])),
+            qml.sample(qml.Projector([1], wires=[1])),
+            qml.sample(qml.Projector([1], wires=[2])),
+        ]
+
+        for node, expected_node in zip(list(g.nodes), expected_nodes):
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:412]:
```suggestion
        for node, expected_node in zip(g.nodes, expected_nodes):
```
Do we need to cast to a list?

File: tests/transforms/test_qcut.py [hunk 8]
```diff
@@ -382,6 +382,87 @@ def test_multiple_observables(self):
...
+            qml.CNOT(wires=[0, 1]),
+            qml.PauliX(wires=[1]),
+            qml.WireCut(wires=[1]),
+            qml.CNOT(wires=[1, 2]),
+            qml.sample(qml.PauliZ(wires=[0])),
+            qml.sample(qml.PauliZ(wires=[1])),
+            qml.sample(qml.PauliZ(wires=[2])),
+        ]
+
+        for node, expected_node in zip(list(g.nodes), expected_nodes):
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:462]:
```suggestion
        for node, expected_node in zip(g.nodes, expected_nodes):
```

File: tests/transforms/test_qcut.py [hunk 9]
```diff
@@ -382,6 +382,87 @@ def test_multiple_observables(self):
...
+        with qml.tape.QuantumTape() as tape:
+            qml.Hadamard(wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.PauliX(wires=1)
+            qml.WireCut(wires=1)
+            qml.CNOT(wires=[1, 2])
+            qml.sample(qml.PauliX(0) @ qml.PauliY(1))
+
+        with pytest.raises(ValueError, match="Sampling from tensor products of observables "):
+            g = qcut.tape_to_graph(tape)
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:431]:
```suggestion
            qcut.tape_to_graph(tape)
```

**Review by trbromley — APPROVED** [review]
Thanks @anthayes92. Please resolve the suggestions, but otherwise approved!

**Pull request merged**